### DISCUSSION TITLE
TangentOperator and PredictionOperator in Marigo mfront

### DISCUSSION THREAD

**Discussion by SCrenSC** [thelfer/MFrontGallery#33] [Q&A]
Hi Thomas,
I tested the Marigo model in the gallery of Mfront. I find that the @PredictionOperator and @TangentOperator blocks are not used since the matrix is calculated in the Integrator block.
I am wondering are there particular reasons to keep these two blocks in the 'Marigo.mfront' template for example? If I remove competently these blocks I will get a message 'umatMarigo.cxx:256:124: required from here
Marigo.mfront:82:6: warning: unused variable 'computeTangentOperator_' '. However, the Mtest could still work.

Thank you,
Sicong

**Comment by thelfer** [comment]
@SCrenSC Are you talking about this implementation: <https://github.com/kuateric/MFrontGallery/blob/master/generic-behaviours/damage/Marigo.mfront> ?

This implementation does define the `@PredictionOperator` and `@TangentOperator` blocks. However, they are only use if the solver request them. By default, `Cast3M` don't use the tangent operator: it uses an accelerated fixe point algorithm in place of the standard Newton method. When using behaviour built with the `Cast3M` interface, `MTest` mimics `Cast3M`' algorithm.

If the `@TangentOperator` block is not defined, we expect the user to compute it in the `@Integrator` block. To check of the constistent a tangent operator has been requested, the user must check the value of the `computeTangentOperator_`. The warning that you report only means that something may go wrong at runtime if the solver does request the consistent tangent operator (it will go wrong because it is not computed by the behaviour).

**Reply by SCrenSC** [reply]
Thank you very much for your quick reply. It is the Marigo.mfront I mentioned. 
May I also ask who will use these 2 blocks among those supported interfaces? I am not familiar with all of them. 
Thanks a lot.

**Reply by thelfer** [reply]
`@PredictionOperator` is only used by `Abaqus/Explicit` and `code_aster`.

`@TangentOperator` is used by most solvers, except `Cast3M` and `FFT`-based solvers, such as `AMITEX_FFTP`.
